### PR TITLE
[SWP-3276] Add Subscription / Queue Listeners Support + refactoring into PubSub package

### PR DESCRIPTION
**Pull request by clemblanco**
This is essentially adding the `sub` aspect to the existing `pub` part of the package. It's relying on the same dependencies and made sense to offer a complete pubsub solution through this package.

The repository will be renamed `laravel-aws-pubsub` once this is merged.

---

**Todo list**
- [x] Implemented the ability to subscribe to events queued on SQS via the `SnsSqs` Queue Connector via listeners instead of jobs
- [x] Ability to define the listeners within a `PubSubServiceProvider` and resolve the `ListenerName` dynamically
- [x] Subject listeners take priority over Topic listeners
- [x] Only rich notification messages are supported, not raw ones
- [x] Queue suffix + prefix are supported
- [x] Covering the Sub aspect with tests
- [x] Polishing the tests
- [x] Refactored the package into a PubSub package offering a complete solution for both the publication (via Broadcasting) and the subscription (using queue Listeners) of server-side events.
- [x] Remove `broadastEvents()` on "Pub part" support as duplicated from Laravel's core
- [x] Add `pubsub:install` command
- [x] Cover it with some tests
- [x] Add `pubsub:make:listener` command
- [x] Cover it with some tests
- [x] Documentation for Sub part
- [x] Documentation unification
- [x] Check Laravel 5.4/5.7 compatibility (only L8 compatible for now)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Changelog
 
-All notable changes to `laravel-sns-broadcaster` will be documented in this file.
+All notable changes to `laravel-aws-pubsub` will be documented in this file.
```
Comment (willthorpe) [inline CHANGELOG.md:3]: TODO later, we will need a new release for this and changelog updated

File: README.md [hunk 2]
```diff
@@ -1,27 +1,41 @@
...
 
-However, we believe this approach of leveraging broadcasting makes sense for a Pub/Sub architecture where an application would like to broadcast an event to the outside world about something that just happened.
+We understand [Broadcasting](https://laravel.com/docs/master/broadcasting) is usually used to "broadcast" your server-side Laravel [Events](https://laravel.com/docs/master/events) over a WebSocket connection to your client-side JavaScript application. However, we believe this approach of leveraging broadcasting makes sense for a Pub/Sub architecture where an application would like to broadcast a server-side event to the outside world about something that just happened.
 
 In this context, "channels" can be assimilated to "topics" on SNS.
 
-## Installation
+**The Sub**
+
+This part is pretty straight forward, we simply have to listen to these messages pushed to an SQS queue and act upon. The only difference here is that we don't use the default Laravel SQS driver as the messages pushed are not following Laravel's classic JSON payload for queued Jobs/Events pushed from a Laravel application. The messages from SNS are simpler.
```
Comment (willthorpe) [inline README.md:18]: This part is pretty straight forward, we simply have to listen to these messages pushed to an SQS queue and act upon **them**. `

or something similar.

File: README.md [hunk 3]
```diff
@@ -1,27 +1,41 @@
...
+
+## Publishing / Broadcasting
+
+### Installation
 
 You can install the package on a Laravel 8+ application via composer:
 
 ```bash
-composer require pod-point/laravel-sns-broadcaster:^0.1
+composer require pod-point/laravel-aws-pubsub
```
Comment (willthorpe) [inline README.md:35]: I think this should be moved outside of the Publishing / Broadcasting header as it is common to both publishing and subscribing.

File: README.md [hunk 4]
```diff
@@ -58,11 +72,9 @@ BROADCAST_DRIVER=sns
 
 Finally, don't forget to enable the [Broadcast Service Provider](https://laravel.com/docs/master/broadcasting#broadcast-service-provider).
 
-## Usage
-
-Here we will simply re-use the power of Laravel Broadcasting, out of the box, with some minor additional functionalities for the Eloquent Model Events.
+---
 
-### Basic Events
+### Usage: Basic Events
```
Comment (willthorpe) [inline README.md:77]: There is a lot of duplication between this section and Usage: Model Broadcasting

I think we need to reorganise this as it feels like I am reading the same stuff again.

File: README.md [hunk 5]
```diff
@@ -238,21 +252,25 @@ In the context of broadcasting, only the following model events can be broadcast
 - `trashed` _if soft delete is enabled_
 - `restored` _if soft delete is enabled_
 
-By default, all of these events are broadcasted, but you can define which events in particular you'd like to broadcast using the `broadcastEvents` method on the model itself:
+By default, all of these events would broadcast, but you can define which events in particular you'd like to broadcast using the `broadcastOn` method on the model itself, just like Laravel suggest it:
```
Comment (willthorpe) [inline README.md:255]: *suggests it.

File: README.md [hunk 6]
```diff
@@ -238,21 +252,25 @@ In the context of broadcasting, only the following model events can be broadcast
...
-    return ['created', 'updated'];
+    return match ($event) {
+        'deleted' => [], // disable broadcasting of the 'deleted' model event
+        default => ['users'],
+    };
 }
 ```
 
-Now only the `created` and `updated` events for this Model will broadcast.
+Now only the `created` and `updated` events for this Model will broadcast (if soft delete is disabled).
```
Comment (willthorpe) [inline README.md:273]: I'm not sure I understand this 🤔  How can it only broadcast created and updated events if you only disable deleted?

I think with the update this doesn't make sense as looking at the old code in the diff it does.

Also on this - the structure of the match() function being an exclude rather than an include is a bit confusing to me but might just be me.

File: README.md [hunk 7]
```diff
@@ -311,6 +329,105 @@ public function broadcastAs($event)
...
+        'prefix' => env('SQS_SNS_PREFIX', 'https://sqs.us-east-1.amazonaws.com/your-account-id'),
+        'queue' => env('SQS_SNS_QUEUE', 'pub-sub'),
+        'suffix' => env('SQS_SNS_SUFFIX'),
+        'region' => env('AWS_DEFAULT_REGION', 'us-east-1'),
+    ],
+    // ...
+],
+```
+
+Once your queue is configured properly, you will need to be able to define which listeners you would like to use for which kind of incoming events. In order to do so, you'll need to create Laravel Listeners and associate the events through a Service Provider the package can create for you.
```
Comment (willthorpe) [inline README.md:354]: We are inconsistent with whether listeners should have a capital or not throughout this file.

File: src/Console/InstallCommand.php [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+     * The console command description.
+     *
+     * @var string
+     */
+    protected $description = 'Install the PubSubEventServiceProvider';
+
+    /**
+     * Execute the console command.
+     *
+     * @return void
```
Comment (willthorpe) [inline src/Console/InstallCommand.php:27]: It can return an int if the PubSubEventServiceProvider already exists.

File: src/Console/InstallCommand.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+     * @param  string  $after
+     * @param  string  $name
+     * @return void
+     */
+    protected function installServiceProviderAfter($after, $name)
+    {
+        if (! Str::contains($appConfig = file_get_contents(config_path('app.php')), 'App\\Providers\\'.$name.'::class')) {
+            file_put_contents(config_path('app.php'), str_replace(
+                'App\\Providers\\'.$after.'::class,',
+                'App\\Providers\\'.$after.'::class,'.PHP_EOL.'        App\\Providers\\'.$name.'::class,',
```
Comment (willthorpe) [inline src/Console/InstallCommand.php:54]: Spacing looks very weird here.

File: src/Pub/Database/Eloquent/BroadcastsEvents.php [hunk 10]
```diff
@@ -1,14 +1,12 @@
 <?php
 
-namespace PodPoint\SnsBroadcaster;
+namespace PodPoint\AwsPubSub\Pub\Database\Eloquent;
 
 use Illuminate\Database\Eloquent\BroadcastsEvents as EloquentBroadcastsEvents;
 
 trait BroadcastsEvents
```
Comment (willthorpe) [inline src/Pub/Database/Eloquent/BroadcastsEvents.php:7]: Is this file necessary? The newBroadcastableModelEvent function looks like a direct copy of the one it inherits from.

File: src/Sub/Queue/Jobs/SqsSnsJob.php [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+     * @param  string  $connectionName
+     * @param  string  $queue
+     * @return void
+     */
+    public function __construct(
+        Container $container,
+        SqsClient $sqs,
+        array $job,
+        $connectionName,
+        $queue
```
Comment (willthorpe) [inline src/Sub/Queue/Jobs/SqsSnsJob.php:28]: ConnectionName and queue have no type here.

File: src/Sub/Queue/Jobs/SqsSnsJob.php [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * Get the underlying raw SQS job.
+     *
+     * @return array
+     */
+    public function getSqsSnsJob()
+    {
+        return $this->job;
+    }
```
Comment (willthorpe) [inline src/Sub/Queue/Jobs/SqsSnsJob.php:87]: This is being flagged as not being used anywhere

File: src/Sub/Queue/SqsSnsQueue.php [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+            return new SqsSnsJob(
+                $this->container, $this->sqs, $response['Messages'][0],
+                $this->connectionName, $queue
+            );
+        }
+    }
+
+    /**
+     * Check wether there is a message pushed from SNS to SQS
```
Comment (willthorpe) [inline src/Sub/Queue/SqsSnsQueue.php:60]: *Whether

File: tests/Console/ListenerMakeCommandTest.php [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+        $this->cleanup();
+    }
+
+    public function tearDown(): void
+    {
+        parent::tearDown();
+
+        $this->cleanup();
+    }
```
Comment (willthorpe) [inline tests/Console/ListenerMakeCommandTest.php:21]: Do we definitely need both of these?

File: tests/Pub/TestClasses/Models/UserWithBroadcastingEventsForSpecificEvents.php [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        switch ($event) {
+            case 'created':
+            case 'trashed':
+            case 'restored':
+            case 'deleted':
+                return [];
+            case 'updated':
+                return ['users'];
+        }
```
Comment (willthorpe) [inline tests/Pub/TestClasses/Models/UserWithBroadcastingEventsForSpecificEvents.php:21]: Not sure if we should change the name of the file to reflect it is only doing Updated?

File: src/Console/InstallCommand.php [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    /**
+     * Install the service provider in the application configuration file.
+     *
+     * @param  string  $after
+     * @param  string  $name
+     * @return void
+     */
+    protected function installServiceProviderAfter($after, $name)
```
Comment (willthorpe) [inline src/Console/InstallCommand.php:49]: Can add string typing to this

File: src/Console/ListenerMakeCommand.php [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+        return __DIR__.'/stubs/listener.stub';
+    }
+
+    /**
+     * Get the default namespace for the class.
+     *
+     * @param  string  $rootNamespace
+     * @return string
+     */
+    protected function getDefaultNamespace($rootNamespace)
```
Comment (willthorpe) [inline src/Console/ListenerMakeCommand.php:51]: Can add string typing to this.

**Review by willthorpe — CHANGES_REQUESTED** [review]
Couple of organisation questions and we should update to use the new laravel broadcasting stuff I guess

File: README.md (hunk 4 above)
Comment (clemblanco) [reply README.md:77]: Myeah I agree...

File: README.md (hunk 6 above)
Comment (clemblanco) [reply README.md:273]: This is referring to the [official documentation](https://laravel.com/docs/8.x/broadcasting#model-broadcasting)

Image 2021-07-28 at 4 53 30 pm

> How can it only broadcast created and updated events if you only disable deleted?
Because it would fallback to use whatever you have defined in your `default`.
This would be similar to have
```php
switch($event) {
    case 'deleted':
        return [];
    default:
        return ['users'];
}
```
Which means for `deleted` we don't define any channel but for any other event we do define some.

Comment (willthorpe) [reply README.md:273]: I think we need to adjust this text then:
`Now only the 'created' and 'updated' events for this Model will broadcast (if soft delete is disabled).`

To something like `Now the soft delete event will not be broadcast`

Comment (willthorpe) [reply README.md:273]: Not very elegant English 😅  but the old text implied that more than just deleted was excluded.

File: src/Console/InstallCommand.php (hunk 9 above)
Comment (clemblanco) [reply src/Console/InstallCommand.php:54]: This is normal, see https://github.com/laravel/jetstream/blob/2.x/src/Console/InstallCommand.php#L474

File: src/Pub/Database/Eloquent/BroadcastsEvents.php (hunk 10 above)
Comment (clemblanco) [reply src/Pub/Database/Eloquent/BroadcastsEvents.php:7]: It's not necessary anymore with Laravel 8.52.

The only thing different in this file was the use of our own `BroadcastableModelEventOccurred` instead of Laravel one.

File: src/Sub/Queue/Jobs/SqsSnsJob.php (hunk 11 above)
Comment (clemblanco) [reply src/Sub/Queue/Jobs/SqsSnsJob.php:28]: We inherit from Laravel's one and Laravel's one doesn't have types either so we cannot add them otherwise it will be marked as "incompatible".

https://github.com/laravel/framework/blob/8.x/src/Illuminate/Queue/Jobs/SqsJob.php#L35

File: src/Sub/Queue/Jobs/SqsSnsJob.php (hunk 12 above)
Comment (clemblanco) [reply src/Sub/Queue/Jobs/SqsSnsJob.php:87]: It's now used in the tests.

File: tests/Console/ListenerMakeCommandTest.php (hunk 14 above)
Comment (clemblanco) [reply tests/Console/ListenerMakeCommandTest.php:21]: Yes, do you understand why?

File: tests/Pub/TestClasses/Models/UserWithBroadcastingEventsForSpecificEvents.php (hunk 15 above)
Comment (clemblanco) [reply tests/Pub/TestClasses/Models/UserWithBroadcastingEventsForSpecificEvents.php:21]: Can do yeah.

File: src/Console/InstallCommand.php (hunk 16 above)
Comment (clemblanco) [reply src/Console/InstallCommand.php:49]: Myeah.

File: src/Console/ListenerMakeCommand.php (hunk 17 above)
Comment (clemblanco) [reply src/Console/ListenerMakeCommand.php:51]: Nope because Laravel's definition of it is not using type hinting.

**Comment by clemblanco** [comment]
@willthorpe ready for another look 👀

**Review by willthorpe — APPROVED** [review]
Noice

There is still a little bit of should Topic and Listener be capitalised in the Readme but it is very very minor.

File: README.md (hunk 6 above)
Comment (clemblanco) [reply README.md]: To be honest this bit of text is now gone and I just point to the official documentation.
